ocaml/typing/typecore.ml

```diff
@@ -645,19 +651,23 @@ let extract_label_names env ty =
 let is_principal ty =
   not !Clflags.principal || get_level ty = generic_level
 
-let has_local_attr loc attrs =
-  match Builtin_attributes.has_local attrs with
-  | Ok l -> l
-  | Error () ->
-     raise(Typetexp.Error(loc, Env.empty, Unsupported_extension Local))
+let unwrap_local_pat spat = match Jane_syntax.Pattern.of_ast spat with
+  | Some (Jpat_local (Lpat_local spat), _) -> Some spat
+  | Some _ | None -> None
 
-let has_local_attr_pat ppat =
-  has_local_attr ppat.ppat_loc ppat.ppat_attributes
+let get_fun_arg_alloc_mode spat =
+  match unwrap_local_pat spat with
+  | Some _ -> Alloc_mode.Local
+  | None   -> Alloc_mode.Global
 
-let has_local_attr_exp pexp =
-  has_local_attr pexp.pexp_loc pexp.pexp_attributes
+let extract_constraint_alloc_mode pexp =
+  match Jane_syntax.Expression.of_ast pexp with
+  | Some (Jexp_local (Lexp_constrain_local pexp), attrs) ->
+      Alloc_mode.Local, {pexp with pexp_attributes = attrs}
```

Comment:
I think this should append, though -- at least, that's the behavior we had historically, since `[@local]` was just one attribute among many.  Does that sounds right to you?

Reply:
Oh I missed that this was *constrain* local.  Then yes, I agree, but also you'll be eliminating this soon anyway.

Reply:
Yep, agreed. I see that you changed this to append, so I'll resolve this comment.